### PR TITLE
CSM: Add method to display a formspec and callbacks

### PR DESCRIPTION
**Pull request by red-001**

### CODE REVIEW

**Comment by nerzhul** [comment]
wrong branch :(

File: builtin/client/register.lua [hunk 1]
```diff
@@ -61,5 +61,6 @@ core.registered_on_sending_chat_messages, core.register_on_sending_chat_messages
 core.registered_on_death, core.register_on_death = make_registration()
 core.registered_on_hp_modification, core.register_on_hp_modification = make_registration()
 core.registered_on_damage_taken, core.register_on_damage_taken = make_registration()
+core.registered_on_formspec_input, core.register_on_formspec_input = make_registration()
```
Comment (rubenwardy) [inline builtin/client/register.lua]: Player receive fields

File: src/script/lua_api/l_client.cpp [hunk 2]
```diff
@@ -59,10 +59,28 @@ int ModApiClient::l_display_chat_message(lua_State *L)
...
+int ModApiClient::l_show_formspec(lua_State *L)
+{
+
+	const char *formname = luaL_checkstring(L, 1);
+	const char *formspec = luaL_checkstring(L, 2);
+
+	ClientEvent event;
+	event.type = CE_SHOW_LOCAL_FORMSPEC;
+	// pointer is required as event is a struct only!
+	// adding a std:string to a struct isn't possible
```
Comment (rubenwardy) [inline src/script/lua_api/l_client.cpp]: I'm not sure about this

File: builtin/client/register.lua (hunk 1 above)
Comment (nerzhul) [inline builtin/client/register.lua]: i'm not use it's logical as there is network reception there

File: src/script/lua_api/l_client.cpp (hunk 2 above)
Comment (nerzhul) [inline src/script/lua_api/l_client.cpp]: it's totally possible, const char* is not possible. I don't know who does this comment but it's surely totally wrong

File: builtin/client/register.lua (hunk 1 above)
Comment (red-001) [inline builtin/client/register.lua]: That name doesn't really make sense for client-sided mods as there is only one player and the fields aren't sent anywhere.

File: src/script/lua_api/l_client.cpp (hunk 2 above)
Comment (red-001) [inline src/script/lua_api/l_client.cpp]: I think I might have accidentally copied it from another part of the code.

Comment (nerzhul) [inline src/script/lua_api/l_client.cpp]: @rubenwardy i will provide a patch on a PR for master to remove that wrong thing, the only problematic thing is the event memory copy here, which could be huge client side, and should use pointers

**Comment by red-001** [comment]
Updated and rebased.

File: src/client.h [hunk 3]
```diff
@@ -175,6 +176,10 @@ struct ClientEvent
 			std::string *formspec;
 			std::string *formname;
 		} show_formspec;
+		struct{
+			const char* formspec;
```
Comment (nerzhul) [inline src/client.h]: no, use std::string (without pointer please)

File: src/game.cpp [hunk 4]
```diff
@@ -3271,6 +3255,12 @@ void Game::processClientEvents(CameraOrientation *cam, float *damage_flash)
 
 			delete(event.show_formspec.formspec);
 			delete(event.show_formspec.formname);
+		} else if (event.type == CE_SHOW_LOCAL_FORMSPEC) {
+			FormspecFormSource *fs_src = new FormspecFormSource(event.show_formspec_local.formspec);
+			std::string name = event.show_formspec_local.formname;
```
Comment (nerzhul) [inline src/game.cpp]: use directly event.show... in LocalFormspecHandler, var creation is not required there

File: src/guiFormSpecMenu.h [hunk 5]
```diff
@@ -547,6 +547,11 @@ class FormspecFormSource: public IFormSource
 		m_formspec = formspec;
 	}
 
+	FormspecFormSource(const char &formspec)
```
Comment (nerzhul) [inline src/guiFormSpecMenu.h]: const std::string& instead of const char &, you only have 1 char here

File: src/client.h (hunk 3 above)
Comment (red-001) [inline src/client.h]: you can't do that, `std::string` inside a union causes errors, at the very least my compiler can't handle it.

File: doc/client_lua_api.txt [hunk 6]
```diff
@@ -710,6 +713,8 @@ Call these functions only at load time!
 * `minetest.after(time, func, ...)`
     * Call the function `func` after `time` seconds, may be fractional
     * Optional: Variable number of arguments that are passed to `func`
+	
+
```
Comment (nerzhul) [inline doc/client_lua_api.txt]: empty lines

File: src/game.cpp [hunk 7]
```diff
@@ -125,12 +125,13 @@ struct TextDestPlayerInventory : public TextDest {
...
 struct LocalFormspecHandler : public TextDest {
 	LocalFormspecHandler();
+
 	LocalFormspecHandler(std::string formname) :
 		m_client(0)
 	{
 		m_formname = formname;
 	}
-
+
```
Comment (nerzhul) [inline src/game.cpp]: unneeded indent

File: src/script/lua_api/l_client.cpp [hunk 8]
```diff
@@ -44,18 +45,56 @@ int ModApiClient::l_get_last_run_mod(lua_State *L)
...
 	getScriptApiBase(L)->setOriginDirect(mod);
-	return 0;
+	lua_pushboolean(L, true);
+	return 1;
 }
 
 // display_chat_message(message)
 int ModApiClient::l_display_chat_message(lua_State *L)
 {
 	NO_MAP_LOCK_REQUIRED;
```
Comment (nerzhul) [inline src/script/lua_api/l_client.cpp]: this is not needed in client

File: builtin/client/init.lua [hunk 9]
```diff
@@ -6,8 +6,18 @@ local commonpath = scriptpath.."common"..DIR_DELIM
 dofile(clientpath .. "register.lua")
 dofile(commonpath .. "after.lua")
 dofile(commonpath .. "chatcommands.lua")
+dofile(clientpath .. "chatcommands.lua")
```
Comment (rubenwardy) [inline builtin/client/init.lua]: mistake?

Comment (nerzhul) [inline builtin/client/init.lua]: hmmm common has registration, and here we need client side definitions but it seems this PR doesn't contain chatcommands.lua, missing chatcommands.lua @red-001  ?

Comment (rubenwardy) [inline builtin/client/init.lua]: there shouldn't be any chatcommands added by this pr

Comment (red-001) [inline builtin/client/init.lua]: just a missing include from a previous  pr, probably lost while rebasing

**Comment by red-001** [comment]
Update.

**Comment by nerzhul** [comment]
merged

**Comment by paramat** [comment]
@red-001 The lack of a description or descriptive commit message is unnacceptable, see https://github.com/minetest/minetest/issues/6736#issuecomment-358156393

**Comment by paramat** [comment]
Where's the 2nd approval?

**Comment by paramat** [comment]
Description should have been added.

**Pull request merged**